### ISSUE TITLE
Stop writing Label Index and Label Offset Table in the index

### ISSUE THREAD

**Issue by codesome** [prometheus/prometheus#14749]
### Proposal

Label Index and Label Offset Table are still being written to the index for backward compatibility. It has been a long time since we stopped reading that in TSDB. 3.0 would be a good time to get rid of those fields. [This](https://github.com/prometheus/prometheus/blob/9a813d5d0f9fd6450f8c6a22f3a049a0f76b8a95/tsdb/index/index.go#L398) and [this](https://github.com/prometheus/prometheus/blob/9a813d5d0f9fd6450f8c6a22f3a049a0f76b8a95/tsdb/index/index.go#L408).

```
┌────────────────────────────┬─────────────────────┐
│ magic(0xBAAAD700) <4b>     │ version(1) <1 byte> │
├────────────────────────────┴─────────────────────┤
│ ┌──────────────────────────────────────────────┐ │
│ │                 Symbol Table                 │ │
│ ├──────────────────────────────────────────────┤ │
│ │                    Series                    │ │
│ ├──────────────────────────────────────────────┤ │
│ │                 Label Index 1                │ │ <- remove from here
│ ├──────────────────────────────────────────────┤ │
│ │                      ...                     │ │ ...
│ ├──────────────────────────────────────────────┤ │
│ │                 Label Index N                │ │ <- remove till here
│ ├──────────────────────────────────────────────┤ │
│ │                   Postings 1                 │ │
│ ├──────────────────────────────────────────────┤ │
│ │                      ...                     │ │
│ ├──────────────────────────────────────────────┤ │
│ │                   Postings N                 │ │
│ ├──────────────────────────────────────────────┤ │
│ │               Label Offset Table             │ │ <- remove this
│ ├──────────────────────────────────────────────┤ │
│ │             Postings Offset Table            │ │
│ ├──────────────────────────────────────────────┤ │
│ │                      TOC                     │ │
│ └──────────────────────────────────────────────┘ │
└──────────────────────────────────────────────────┘
```

**Comment by jan--f** [comment]
@codesome I added the `breaking` label, but not sure if that's accurate. Iiuc with this change the resulting tsdb can no longer be opened by older Prometheus versions?

**Comment by gouthamve** [comment]
I don't think we should block rolling back from 3.0 to 2.54. The right way to solve this is to do the following:

1. Add a new version of the index with the fields removed.
2. Continue to write older version indexes for N number of releases
3. Cut over to new version after N releases (once 3.x has been sufficiently stable) --> this is the point where you can't go back to 2.54


But yeah, maybe we should add the new version to the 3.0 release goals.

**Comment by codesome** [comment]
Alternate idea: release a 2.55 with the only diff from 2.54 being the code to support rollback from 2.55. 2.55 release could add rollback code for other features as well (if this issue is the only one, then maybe a new release for just this is not required)

This way we could start writing new index version starting 3.0, and 2.55 can still work with it. This removed fields are not read for last many versions anyway.

**Comment by codesome** [comment]
I have a wild unconventional idea: keep the index version as V2 but don't write these fields anymore, and keep the TOC structure intact. The 2.x version from last couple of years (or more? need to check) will still be able to read the index that way, because they don't access those sections or check TOC for those sections. This will avoid doing breaking change in 3.X (X >= 1).

With 3.0 branch merged into main branch, it is not easy to do a 2.X right now. Maybe we backport the backward compatibility in existing 2.X releases with a patch release, which is not ideal.

I feel it is a good compromise to say with 3.0 you cannot downgrade to before 2.Z (Z to be determined).

Or is it more acceptable to say you cannot downgrade from >=3.1 to 2.Y and you can only downgrade from >=3.1 to 3.0?

**Comment by codesome** [comment]
So, [2.15.0 / 2019-12-23](https://github.com/prometheus/prometheus/releases/tag/v2.15.0) is the release after which we stopped reading these fields (~5 years ago). So I feel keeping the index version as V2 and stop writing it should be okay, with a message that after you upgrade to 3.0, you cannot downgrade below 2.15.0, should be acceptable

**Comment by gouthamve** [comment]
> So I feel keeping the index version as V2 and stop writing it should be okay, with a message that after you upgrade to 3.0, you cannot downgrade below 2.15.0, should be acceptable

You're assuming that Prometheus is the only one reading this. TSDB was an independent repo and library and TSDB blocks are used in Mimir, Thanos and maybe other projects that we don't know about. I would like to follow common engineering processes here and bump the version number if we're changing the format.

**Comment by codesome** [comment]
Step 1: Adding backward compatibility to v2.55 https://github.com/prometheus/prometheus/pull/14934

**Comment by codesome** [comment]
If we do this change in 3.0, users will only be able to downgrade to 2.55 and not before. This is not a friendly scenario. So I will do this change in some future version, maybe 3.3 or 3.4. That should give a good 4-5 releases to downgrade to.

**Comment by gavine** [comment]
Hi @codesome ,

Is downgrading from v3.x to a pre-v2.55.0 version completely unsupported, or is it technically possible but with the loss of FormatV3 data after rollback?

The V3 migration guide states, '... you will only be able to downgrade to v2.55, not lower, without losing your TSDB persistent data.' This implies (to me) the latter. However, the code (https://github.com/prometheus/prometheus/blob/main/tsdb/db.go#L1600-L1615) suggests that the downgrade may fail, meaning it is not supported at all.

Thanks!

**Comment by krajorama** [comment]
Hello from the bug scrub!

Seems to be in discussion and depends on LTS related developments, keeping open.

**Comment by bwplotka** [comment]
Potentially related discussion on persistent upgrades https://github.com/prometheus/proposals/pull/40

**Comment by bwplotka** [comment]
We decided to backport compatibility support to the current LTS (2.53) https://github.com/prometheus/prometheus/pull/16762

Once 2.53 will have it (likely 2.53.5), we can perform the removal 👍🏽

**Comment by bwplotka** [comment]
To answer:
 https://github.com/prometheus/prometheus/issues/14749#issuecomment-2632524838

When I tested the "incompatible" downgrade, Prometheus simply fails to start. To recover one would need to delete the problematic tsdb blocks. Is that helpful @gavine ?

**Comment by bwplotka** [comment]
With 2.53.5  this change is now unblocked (fields can be removed)

**Comment by pipiland2612** [comment]
Hi @bwplotka, Given that backward compatibility support for the upcoming index v3 has been backported to Prometheus 2.53.5 (as confirmed in recent discussions), I believe we can now safely proceed with removing the Label Index and Label Offset Table sections from the TSDB index. If you agree, I'd like to move forward and implement this change. Please let me know your thoughts!

**Comment by bwplotka** [comment]
Yes, I believe we can do this now 👍 Thanks for helping!
